neps/nep-0366.md

```diff
@@ -0,0 +1,148 @@
+    /// After this action is processed it will increment.
+    nonce: Nonce,
+    /// Signature of the originating user signing `DelegateActionMessage` formed out of the data in the `DelegateAction`.
+    signature: Signature,
+}
+```
+
+ Supporting a batch of `actions` means `DelegateAction` can initiate a complex steps like creating a new account and transferring funds, deploying a contract and executing an initialization function.
+
+***Validation***:
+To ensure that `DelegateAction` is correct, on receival the verification of signature is done: `verify_signature(hash(message), signer_pk, signature)`.
+
+The `message` is formed in the next format and must be signed by the user:
+```rust
+struct DelegateActionMessage {
+    /// If not None, should match the `predecessor_id` that have created `DelegateAction`.
+    sender_id: Option<AccountId>,
+    /// Matching the given account_id.
+    signer_id: AccountId,
+    /// Matching the `signer_pk` from `DelegateAction`.
+    public_key: PublicKey,
+    /// Nonce for the given `public_key` from `DelegateAction`.
+    nonce: Nonce,
+    /// Matching the `receiver_id` from `DelegateAction`.
+    receiver_id: AccountId,
+    /// Block hash to ensure validity.
+    /// Actions matching `actions` from `DelegateAction`.
+    actions: Vec<Action>,
+}
+```
+
+The next set of security concerns are addressed by this format:
+ - If format matches `Transaction`, the relayer can just send it directly, not receiving payment.
+ - Because set of actions can include pay back to the relayer (for example by paying in FT), the `sender_id` is added directly into the message to ensure that nobody else can send this message.
+ - `nonce` is included to ensure that this message can't be replayed again.
+ - `public_key` and `signer_id` are needed to ensure the on the right account, work across rotating keys and fetch correct `nonce`.
+
+The permissions are verified based on kind `AccessKeyPermission` of `signer_pk`:
+ - `AccessKeyPermission::FullAccess`, all actions are allowed.
+ - `AccessKeyPermission::FunctionCall`, only a single `FunctionCall` action is allowed in `actions`. 
+    - `DelegateAction.receiver_id` must match to the `account[public_key].receiver_id`
+    - `DelegateAction.actions[0].method_name` must be in the `account[public_key].method_names`
+
+***Outcomes***:
+- If `signature` matches receiver's accounts `signer_pk`, new receipt is created from this account with set of `ActionReceipt { receiver_id, action }` for each item in `actions`. 
+
+#### Errors
+
+**Validation Error**
+- If `signer_pk` does not exist for the given account, the following error will be returned
+```rust
+/// Signer key for delegate action is missing from given account
+DelegateActionSignerDoesNotExist
+```
+
+- If `nonce` does not exist match `signer_pk` for  `receiver_id`, the following error will be returned
+```rust
+/// Nonce must be account[signer_pk].nonce + 1
+DelegateActionInvalidNonce
```

Comment:
Why do we enforce `+1` here? That is not how access key works in other cases

Reply:
Can you share a reference about how nonces work today? I can't find it in nomicon.